lib/ide/utils/src/handle.rs

```diff
@@ -0,0 +1,268 @@
+
+/// A type that can provide `Weak` handle to a value under `RefCell`.
+pub trait IsWeakHandle: Clone {
+    /// Type of the data stored.
+    type Data;
+
+    /// Obtain weak handle to the data.
+    fn weak_handle(&self) -> WeakHandle<Self::Data>;
+
+    /// Obtain strong handle to the data.
+    fn upgrade(&self) -> Option<StrongHandle<Self::Data>> {
+        self.weak_handle().upgrade()
+    }
+
+    /// Returns `Future` that shall try calling `Cb` with the access to mutably
+    /// borrowed data.
+    fn with<Cb,R> (&self,cb:Cb) -> WeakWith<Self,Cb>
+        where Cb: FnOnce(&mut Self::Data) -> R {
+        WeakWith::new(self.clone(),cb)
+    }
+}
+
+/// Clones the value loaded from the handle.
+pub async fn try_load<Data: Clone>
+(handle:&impl IsWeakHandle<Data=Data>) -> Result<Data> {
+    handle.with(|data| data.clone()).await
+}
+
+/// Stores the value into the handle.
+pub async fn try_store<Data>
+(handle:&impl IsWeakHandle<Data=Data>, value:Data) -> Result<()> {
+    handle.with(|data| *data = value).await
+}
+
+impl<T> IsWeakHandle for WeakHandle<T> {
+    type Data = T;
+    fn weak_handle(&self) -> WeakHandle<Self::Data> {
+        self.clone()
+    }
+}
+
+
+
+// ==================
+// == StrongHandle ==
+// ==================
+
+/// A type that can provide a strong (`Rc`) handle to a value under `RefCell`.
+pub trait IsStrongHandle: Clone {
+    /// Type of the data stored.
+    type Data;
+
+    /// Obtain strong (`Rc`) handle to the data.
+    fn strong_handle(&self) -> StrongHandle<Self::Data>;
+
+    /// Obtain weak handle to the data.
+    fn downgrade(&self) -> WeakHandle<Self::Data> {
+        Rc::downgrade(&self.strong_handle())
+    }
+
+    /// Returns `Future` that shall call `Cb` with the access to mutably
+    /// borrowed data.
+    fn with<Cb,R> (&self, cb:Cb) -> StrongWith<Self,Cb>
+    where Cb: FnOnce(&mut Self::Data) -> R {
+        StrongWith::new(self.clone(),cb)
+    }
+}
+
+/// Clones the value loaded from the handle.
+pub async fn load<Data: Clone>(handle:&impl IsStrongHandle<Data=Data>) -> Data {
+    handle.with(|data| data.clone()).await
+}
+
+/// Stores the value into the handle.
+pub async fn store<Data>(handle:&impl IsStrongHandle<Data=Data>, value:Data) {
+    handle.with(|data| *data = value).await
+}
+
+impl<T> IsStrongHandle for StrongHandle<T> {
+    type Data = T;
+    fn strong_handle(&self) -> StrongHandle<Self::Data> {
+        self.clone()
+    }
+}
+
+
+
+// ==============
+// == WeakWith ==
+// ==============
+
+/// Future that shall call `cb` with the data under `handle`.
+#[derive(Debug)]
+pub struct WeakWith<Handle,Cb> {
+    handle : Handle,
+    cb     : Option<Cb>,
+}
+
+impl<Handle,Cb> WeakWith<Handle,Cb> {
+    /// Creates the value.
+    pub fn new<Data,R>
+    ( handle : Handle
+    , cb     : Cb)
+    -> WeakWith<Handle,Cb>
+    where Handle : IsWeakHandle<Data = Data>,
+          Cb     : FnOnce(&mut Data) -> R, {
+        WeakWith {handle,cb:Some(cb)}
+    }
```

Comment:
This definition makes the type inference more complex than it sohuld be. Moreover, it is hard to read `Data = Data`. Why not (assuming using `HasContent` as proposed above):
```rust
pub fn new<T> (handle:Handle, callback:Callback) -> WeakWith<Handle,Callback>
where Handle   : IsWeakHandle,
      Callback : FnOnce(&mut Content<Handle>) -> R {
    WeakWith {handle,callback:Some(callback)}
}
```

Reply:
For me, the @mwu-tow signature is better, because it's pattern common in Rust (like all the operations on Iterators are defined)
Yeah, `Data = Data` is a little verbose. It is a drawback? For me it's not.

Reply:
simplified, as the `new` does not really care about Callback type, only the `Future` impl

Reply:
`Data = Data` looks funny but imho it's clear once you saw it once — I thought about renaming one of them… but then why should we came up with different names for the very same entity.

(but code is gone anyway)